src/ucp/core/ucp_device.c

```diff
@@ -90,87 +154,286 @@ ucp_device_mem_list_params_check(const ucp_device_mem_list_params_t *params)
-                              "ucp_device_mem_list_handle_t", &mem);
+    /* Allocate handle on the same device memory */
+    num_uct_eps  = i;
+    handle_size *= params->num_elements;
+    handle_size += sizeof(*handle);
+    status       = ucp_mem_do_alloc(ep->worker->context, NULL, handle_size,
+                                    UCT_MD_MEM_ACCESS_LOCAL_READ |
+                                            UCT_MD_MEM_ACCESS_LOCAL_WRITE,
+                                    mem_type, local_sys_dev,
+                                    "ucp_device_mem_list_handle_t", mem);
     if (status != UCS_OK) {
         ucs_error("failed to allocate ucp_device_mem_list_handle_t: %s",
                   ucs_status_string(status));
         return status;
     }
 
-    ucp_memory_detect(ep->worker->context, mem.address, mem.length, &mem_info);
-    if (mem_info.sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN) {
-        ucs_error("failed to detect sys_dev for mem list handle: %s",
-                  ucs_status_string(status));
-        status = UCS_ERR_NO_DEVICE;
+    /* Populate the handle */
+    handle = ucs_alloc_on_stack(handle_size, "host_device_mem_list_handle");
+    if (handle == NULL) {
+        status = UCS_ERR_NO_MEMORY;
         goto err;
     }
 
-    /* TODO actual lane lookup and population of handle */
-    memset(&host_handle, 0, sizeof(host_handle));
-    host_handle.version = UCP_DEVICE_MEM_LIST_VERSION_V1;
+    handle->version            = UCP_DEVICE_MEM_LIST_VERSION_V1;
+    handle->proto_idx          = 0;
+    handle->num_uct_eps        = num_uct_eps;
+    handle->mem_list_length    = params->num_elements;
+    handle->mem_list_offset[0] = 0;
 
-    ucp_mem_type_unpack(ep->worker, mem.address, &host_handle,
-                        sizeof(host_handle), UCS_MEMORY_TYPE_CUDA);
+    /* Populate lane specific parameters */
+    for (i = 0; i < num_uct_eps; i++) {
+        handle->mem_list_offset[i + 1] = uct_elem_size[i] +
+                                         handle->mem_list_offset[i];
+        status = uct_ep_get_device_ep(ucp_ep_get_lane(ep, lanes[i]),
+                                      &handle->uct_device_eps[i]);
+        if (status != UCS_OK) {
+            ucs_error("failed to get device_ep for lane=%u", lanes[i]);
+            /* TODO: actually fail */
+        }
+    }
 
-    ucs_spin_lock(&ucp_device_handle_hash_lock);
-    iter = kh_put(ucp_device_handle_allocs, &ucp_device_handle_hash,
-                  mem.address, &ret);
-    if (ret == UCS_KH_PUT_FAILED) {
-        ucs_error("failed to hash handle=%p", mem.address);
-        goto err_unlock;
-    } else if (ret == UCS_KH_PUT_KEY_PRESENT) {
-        ucs_error("handle=%p already found in hash", mem.address);
-        goto err_unlock;
+    /* Populate element specific parameters */
+    ucp_element = params->elements;
+    uct_element = UCS_PTR_BYTE_OFFSET(handle, sizeof(*handle));
+    for (i = 0; i < params->num_elements; i++) {
+        for (j = 0; j < num_uct_eps; j++) {
```

Comment:
Maybe I got it wrong, but I think we need to reverse the order, i=uct_ep_idx and j=elem_idx.
uct device API expect a pointer to a list of elements.